crates/runtime/src/uuid.rs

```diff
@@ -0,0 +1,234 @@
+// and 1 Jan 1970 (Unix epoch), in units of 100 nanoseconds.
+const G1582NS100: u64 = 122_192_928_000_000_000;
+
+/// Parse a v1 UUID into its Producer, Clock, and Flags.
+/// If it's not a V1 UUID then `parse_uuid` returns None.
+#[inline]
+pub fn parse_uuid(u: uuid::Uuid) -> Option<(Producer, Clock, Flags)> {
+    if u.get_version_num() != 1 {
+        return None;
+    }
+    let (c_low, c_mid, c_high, seq_node_id) = u.as_fields();
+
+    let clock = (c_low as u64) << 4          // Clock low bits.
+            | (c_mid as u64) << 36                  // Clock middle bits.
+            | (c_high as u64) << 52                 // Clock high bits.
+            | ((seq_node_id[0] as u64) >> 2) & 0xf; // High 4 bits of sequence number.
+
+    // 6 bytes of big-endian node ID.
+    let producer: [u8; 6] = seq_node_id[2..8].try_into().unwrap();
+
+    let flags = ((seq_node_id[0] as u16) & 0x3) << 8 // High 2 bits of flags.
+            | (seq_node_id[1] as u16); // Low 8 bits of flags.
+
+    Some((Producer(producer), Clock(clock), Flags(flags)))
+}
+
+/// Build a V1 UUID from a Producer, Clock, and Flags.
+#[inline]
+pub fn build_uuid(p: Producer, c: Clock, f: Flags) -> uuid::Uuid {
+    assert!(f.0 <= 0x3ff, "only 10 low bits may be used for flags");
+
+    let p1 = ((c.0 >> 4) as u32).to_be_bytes(); // Clock low bits.
+    let p2 = ((c.0 >> 36) as u16).to_be_bytes(); // Clock middle bits.
+    let p3 = ((c.0 >> 52) as u16 | 0x1000).to_be_bytes(); // Clock high bits + version 1.
+    let p4 = (((c.0 << 10) as u16) & 0x3c00 | f.0 | 0x8000).to_be_bytes(); // Clock sequence + flags + variant 1.
+
+    uuid::Uuid::from_bytes([
+        p1[0], p1[1], p1[2], p1[3], p2[0], p2[1], p3[0], p3[1], p4[0], p4[1], p.0[0], p.0[1],
+        p.0[2], p.0[3], p.0[4], p.0[5],
+    ])
+}
+
+/// Build a flow::UuidParts from a Producer, Clock, and Flags.
+pub fn build_uuid_parts(p: Producer, c: Clock, f: Flags) -> flow::UuidParts {
+    flow::UuidParts {
+        clock: c.0,
+        node: (p.0[0] as u64) << 56
+            | (p.0[1] as u64) << 48
+            | (p.0[2] as u64) << 40
+            | (p.0[3] as u64) << 32
+            | (p.0[4] as u64) << 24
+            | (p.0[5] as u64) << 16
+            | (f.0 as u64),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_clock_lifecycle() {
+        let mut c = Clock::UNIX_EPOCH;
+        assert_eq!(c.0, 0x1b21dd2138140000);
+
+        // Each tick increments the clock.
+        c.tick();
+        c.tick();
+        assert_eq!(c.0, 0x1b21dd2138140002);
+
+        // Updates take the maximum value of the observed Clocks (Clock is monotonic).
+        c.update(Clock::from_unix(10, 0));
+        assert_eq!(c.0, 0x1b21dd2197721000);
+        c.update(Clock::from_unix(5, 0));
+        assert_eq!(c.0, 0x1b21dd2197721000); // Not changed.
+    }
+
+    #[test]
+    fn test_parse_and_build_with_fixture() {
+        let u1 = uuid::Uuid::parse_str("9f2952f3-c6a3-11ea-8802-080607050309").unwrap();
```

Comment:
placeholder: it might be nice to have a corresponding Go test that parses the same uuid string and asserts the same results

Reply:
https://github.com/estuary/flow/blob/8bb5f32aaf06dadf0809b1e571ef40c637b72582/go/protocols/flow/document_extensions_test.go#L35